### PR TITLE
Replace linter and formatter with ruff and mypy

### LINKED ISSUE #364: ci: Modernize the CI tools
It would be good to follow the CI tools in [scikit-build-core](https://github.com/scikit-build/scikit-build-core) to keep up with the modern tools, currently:
- `black` -> `ruff-format`
- `flake8`, `isort`, (maybe also `pydocstyle`) -> `ruff`

### PR DESCRIPTION
**Pull request by lan496**
Closes https://github.com/spglib/spglib/issues/364

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (lan496) [inline .pre-commit-config.yaml]: `database/*.py` is not actively maintained now

File: pyproject.toml
Comment (lan496) [inline pyproject.toml:199]: mypy is not so smart to pass
```
d: dict | None
assert(d is not None)
d.keys()
```

File: pyproject.toml
Comment (LecrisUT) [inline pyproject.toml]: By default `ruff` does not include any linting. It should include something like:
```toml
[tool.ruff.lint]
extend-select = [
    "F",           # pyflakes
    "W",           # pycodestyle-warnings
    "E",           # pycodestyle-errors
    "C90",         # mccabe
    "I",           # isort
    "N",           # pep8-naming
    "D",           # pydocstyle
    "UP",          # pyupgrade
    "YTT",         # flake8-2020
    "ANN",         # flake8-annotations
    "S",           # flake8-bandit
    "BLE",         # flake8-blind-except
    "FBT",         # flake8-boolean-trap
    "B",           # flake8-bugbear
    "A",           # flake8-builtins
    "COM",         # flake8-comma
    "C4",          # flake8-comprehensions
    "T10",         # flake8-debugger
    "EM",          # flake8-errmsg
    "FA",          # flake8-future-annotations
    "ISC",         # flake8-implicit-str-concat
    "ICN",         # flake8-import-conventions
    "G",           # flake8-logging-format
    "PIE",         # flake8-pie
    "T20",         # flake8-print
    "PYI",         # flake8-pyi
    "PT",          # flake8-pytest-style
    "Q",           # flake8-quotes
    "RSE",         # flake8-raise
    "RET",         # flake8-return
    "SLF",         # flake8-self
    "SIM",         # flake8-simplify
    "TID",         # flake8-tidy-imports
    "TCH",         # flake8-type-checking
    "ARG",         # flake8-unused-arguments
    "PTH",         # flake8-use-pathlib
    "TD",          # flake8-todos
    "ERA",         # eradicate
    "PD",          # pandas-vet
    "PGH",         # pygrep-hooks
    "PL",          # pylint
    "TRY",         # tryceratops
    "FLY",         # flynt
    "NPY",         # NumPy
    "PERF",        # perflint
    "FURB",        # refurb
    "RUF",         # Ruff-specific
]
```

File: test/functional/python/test_spglib.py
Comment (lan496) [inline test/functional/python/test_spglib.py:5]: `pip install types-PyYAML` did not work somehow.

File: python/spglib/spglib.py
Comment (LecrisUT) [inline python/spglib/spglib.py]: Why the ignore? Also should scope ignores

**Review by LecrisUT — CHANGES_REQUESTED** [review]
Also, just make sure when you apply `pre-commit` it's a separate commit, and we don't squash commit for this PR

File: pyproject.toml
Comment (LecrisUT) [reply pyproject.toml:199]: It did on various projects :thinking:. I don't see anything special in the configure either:
```toml
[tool.mypy]
files = ["src", "test"]
mypy_path = ["$MYPY_CONFIG_FILE_DIR/src"]
python_version = "3.12"
warn_unused_configs = true
show_error_codes = true
enable_error_code = ["ignore-without-code", "redundant-expr", "truthy-bool"]
strict = true
disallow_untyped_defs = false
disallow_incomplete_defs = false

[[tool.mypy.overrides]]
module = ["fmf_jinja.*"]
disallow_untyped_defs = true
disallow_incomplete_defs = true
```

File: test/functional/python/test_spglib.py
Comment (LecrisUT) [reply test/functional/python/test_spglib.py:5]: It has to be in `pre-commit-config.yaml`

File: .pre-commit-config.yaml
Comment (LecrisUT) [reply .pre-commit-config.yaml]: Better to move the exclude to `pyproject.toml`

File: .pre-commit-config.yaml
Comment (LecrisUT) [inline .pre-commit-config.yaml]:
```suggestion
      - id: ruff
        args: ["--fix", "--show-fixes"]
      - id: ruff-format
```
Don't think `type_or`, and `--exit-non-zero-on-fix` does anything in this case.

File: .pre-commit-config.yaml
Comment (lan496) [reply .pre-commit-config.yaml]: Without it, `pre-commit run --all` complains about `dataset/ptg2symbol.py`. git-commit and `pre-commit run --all` seem to use different settings.

Comment (LecrisUT) [reply .pre-commit-config.yaml]: Even when you set [`exclude`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-exclude)? I'll have to try that locally

File: pyproject.toml
Comment (lan496) [reply pyproject.toml:199]: Do you recommend these configs that you write?

File: test/functional/python/test_spglib.py
Comment (lan496) [reply test/functional/python/test_spglib.py:5]: Thanks!

File: pyproject.toml
Comment (LecrisUT) [reply pyproject.toml:199]: Not very experienced with mypy, but that's what I have on one project, and I think I've taken from `scikit-build-core` and `tmt`. So I am not sure if some of them are redundant, but otherwise, looking at the options set, they make sense to have. Worth trying them out as a starting point anyway

**Comment by LecrisUT** [comment]
Also worth referencing the `scienitific-python` configuration: https://github.com/scientific-python/cookie/blob/main/pyproject.toml

File: python/spglib/spglib.py
Comment (lan496) [reply python/spglib/spglib.py]: Without ignoring, mypy says
```
error: Module "spglib" has no attribute "_spglib"  [attr-defined]
```

Comment (LecrisUT) [reply python/spglib/spglib.py]: Yeah, but that's a genuine issue:
- We should be using relative paths there
- We should be providing a `_spglib.pyi` to type-hint the C interface, even if only internally

File: pyproject.toml
Comment (lan496) [reply pyproject.toml]: Flake8's rules seem to be enabled by default. I've added isort's rules.

File: .pre-commit-config.yaml
Comment (lan496) [reply .pre-commit-config.yaml]: Yes, `pre-commit run --all` ignores it...

File: pyproject.toml
Comment (LecrisUT) [reply pyproject.toml]: Only `E` iirc. Best to be extensive. Here's a list of rules: https://docs.astral.sh/ruff/rules/

I would encourage to keep all of them, but comment out those that are not ready to be added.

File: pyproject.toml
Comment (lan496) [reply pyproject.toml:199]: The strict mode is too strict now. But, I'll add other flags because they looks useful.

File: python/spglib/spglib.py
Comment (lan496) [reply python/spglib/spglib.py]: The relative path did not persuade mypy
```
    from .spglib import _spglib as spg
# -> Module "spglib.spglib" has no attribute "_spglib"  [attr-defined]
```

File: pyproject.toml
Comment (LecrisUT) [reply pyproject.toml:199]: I would add `strict` as a comment with a todo/open PR to address it.

Comment (lan496) [reply pyproject.toml:199]: Ok, I will leave todo as comment.

**Comment by LecrisUT** [comment]
Please split the "add mypy/ruff configuration" and "apply mypy/ruff” as separate commits in 31b2842840989df073e3bd35342c7cbdae8b90f2. You can run `git commit --no-verify` to skip the pre-commit

**Review by LecrisUT — COMMENTED** [review]
:+1: for the current setup as an initial implementation. Mind if I rebase your commits a bit before merging?

File: .pre-commit-config.yaml
Comment (LecrisUT) [inline .pre-commit-config.yaml]: The size of the python bindings is very small, so might as well run mypy on all sources defined in `pyproject.toml`, i.e. running
```console
$ mypy --ignore-missing-imports --scripts-are-modules
```

**Comment by lan496** [comment]
@LecrisUT  No problem. I'll leave the rest to you.

**Review by LecrisUT — APPROVED** [review]

**Comment by LecrisUT** [comment]
Sorry, had to change the committer because I've split the commits. Btw, can you add `docs/readthedocs.org`, and `testing-farm:fedora-39` (and later `fedora-38` as well) to the required workflows to pass? Otherwise if we use `auto-merge` now it would keep those checks dangling.

Also make sure to merge this with a merge commit or rebase to keep the history

**Comment by lan496** [comment]
Thanks to tidy commits. I've updated the branch protection rule. Is it OK?
image

**Comment by LecrisUT** [comment]
Yep, that looks good. I'll bother you again later when I enable Fedora38

**Pull request merged**
